daffodil-runtime1/src/main/scala/org/apache/daffodil/udf/UserDefinedFunctionService.scala

```diff
@@ -0,0 +1,185 @@
+        log(LogLevel.Warning, wstr)
+      }
+
+      if (missingAnnotations.isEmpty && nonUserDefinedFunction.isEmpty) {
+        providerfc.map { fc =>
+          val fcClassName = fc.getName
+          val fInfo = fc.getDeclaredAnnotation(classUserDefinedFunctionInfo)
+          val fns = fInfo.namespaceURI()
+          val fname = fInfo.name()
+          if (List(fns, fname).exists { p => Option(p).isEmpty || p.trim.isEmpty }) {
+            val wstr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}." +
+              (if (Option(fns).isEmpty || fns.trim.isEmpty) " Annotation namespace field is empty or invalid." else "") +
+              (if (Option(fname).isEmpty || fname.trim.isEmpty) " Annotation namespace field is empty or invalid." else "")
+            log(LogLevel.Warning, wstr)
+          }
+          val key = s"$fns:$fname"
+          if (functionProviderLookup.contains(key)) {
+            val estr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}. Duplicate $key found."
+            log(LogLevel.Error, estr)
+          } else {
+            functionProviderLookup += (key -> provider)
+            functionClassesLookup += (key -> fc)
+          }
+        }
+      }
+    }
+  }
+
+  if (functionClassesLookup.isEmpty) {
+    log(LogLevel.Warning, "No user defined functions found. Check that UDF JARs are on classpath and that they are properly registerable by ServiceLoader.")
+    log(LogLevel.Info, s"Current classpath locations:\n${currentClassPath}")
+  }
+
+  def getUserDefinedFunctionClasses = functionClassesLookup.values.toArray
+
+  lazy val allFunctionClasses: String = getUserDefinedFunctionClasses.map {
+    c => s"[${c.getName()} => ${c.getAnnotation(classUserDefinedFunctionInfo)}]"
+  }.mkString("\n")
+
+  /**
+   * Returns an initialized UserDefinedFunction object from the UserDefinedFunctionProvider
+   * based on namespaceURI and function name from the schema
+   *
+   * @return Optional Tuple of initialized UserDefinedFunction, evaluate Method, array of
+   * the evaluate's parameter types and evaluate's return type
+   */
+  def lookupUDFCallingInfo(namespaceURI: String, fname: String): Option[UdfCallingInfo] = {
+    val udfid = s"$namespaceURI:$fname"
+
+    val udfFunctionCallingInfo = InitializedUserDefinedFunctionsCache.getOrElse(udfid, {
+      val provider = functionProviderLookup.get(udfid)
+      val fClass = functionClassesLookup.get(udfid)
+      if (!List(provider, fClass).exists(_.isEmpty)) {
+        val udf = provider.get.lookupInitializedUserDefinedFunction(namespaceURI, fname)
+        val expectedUdfClass = fClass.get
+
+        if (udf != null) {
+          val actualUdfClass = udf.getClass
+          if (actualUdfClass != expectedUdfClass) {
+            log(LogLevel.Error, s"User Defined Function Class Mismatch: $udfid. Expected: $expectedUdfClass Actual: $actualUdfClass")
```

Comment:
Most all your logging here is actually user messaging. So Internationalization applies here as to any SDE/SDW. 

Scala's string-interpolation stuff is really actually a pretty bad idea in any situation that involves strings that are user/system messaging. It completely defeats internationalization.